### PR TITLE
Broker

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: src/Broker/Broker.c [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    size_t packet_size = Packetizer_packetize(&packetizer_data,
+                                              Packetizer_PacketType_Telemetry,
+                                              0,
+                                              (uint16_t)interface,
+                                              packetizer_buffer,
+                                              SPACE_PACKET_PRIMARY_HEADER_SIZE,
+                                              length);
+
+    // sent to driver
+    enum SystemBus bus_id = port_to_bus_map[interface];
```
Comment (Glamhoth) [inline src/Broker/Broker.c:37]: const

File: src/Broker/Broker.c [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    void* driver_private_data = bus_to_driver_private_data[bus_id];
+    driver_send_function send_function = bus_to_driver_send_function[bus_id];
+    send_function(driver_private_data, packetizer_buffer, packet_size);
+
+    Broker_release_lock();
+}
+
+// this shall be called by driver
+void
+Broker_receive_packet(uint8_t* data, size_t size)
```
Comment (Glamhoth) [inline src/Broker/Broker.c:47]: const args

File: src/Broker/Broker.c [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+{
+    Broker_acquire_lock();
+    // mutex
+    uint16_t source;
+    uint16_t destination;
+    size_t data_offset;
+    size_t data_size;
+    int32_t error_code;
+    bool success = Packetizer_depacketize(&packetizer_data,
```
Comment (Glamhoth) [inline src/Broker/Broker.c:56]: const

File: src/Broker/Broker.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+extern void Broker_release_lock();
+
+void
+Broker_initialize()
+{
+    Packetizer_init(&packetizer_data);
+}
+
+void
+Broker_deliver_request(enum RemoteInterface interface, uint8_t* data, size_t length)
```
Comment (Glamhoth) [inline src/Broker/Broker.c:23]: const args

File: src/Broker/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    PUBLIC      Broker.h)
+
+target_include_directories(Broker
+  PUBLIC      ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_link_libraries(Broker
+  PRIVATE     common_build_options)
+
+target_link_libraries(Broker
+  PUBLIC      TASTE::Mocks)
```
Comment (Glamhoth) [inline src/Broker/CMakeLists.txt:15]: If this library is used only by the Broker implementation, then this should be marked as `PRIVATE`

File: src/Broker/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+target_link_libraries(Broker
```
Comment (Glamhoth) [inline src/Broker/CMakeLists.txt:17]: All libraries can/should be listed in one `target_link_libraries` call

File: src/Common/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    PRIVATE     DriverHelper.c
+    PUBLIC      DriverHelper.h)
+
+target_include_directories(Common
+    PUBLIC      ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_link_libraries(Common
+    PRIVATE     common_build_options)
+
+target_link_libraries(Common
```
Comment (Glamhoth) [inline src/Common/CMakeLists.txt:14]: All libraries can/should be listed in one `target_link_libraries` call

File: src/Common/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  PUBLIC      TASTE::Mocks)
```
Comment (Glamhoth) [inline src/Common/CMakeLists.txt:15]: `PRIVATE`

File: src/Common/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+add_format_target(Common)
+
+add_library(TASTE::Common ALIAS Common)
```
Comment (Glamhoth) [inline src/Common/CMakeLists.txt:19]: Maybe `TASTE::RuntimeCommon` to avoid possible name clash?

File: src/Common/DriverHelper.c [hunk 10]
```diff
@@ -0,0 +1,29 @@
+#include "DriverHelper.h"
+
+void*
+get_device_configuration(enum SystemDevice device_id)
```
Comment (Glamhoth) [inline src/Common/DriverHelper.c:4]: const arg

File: src/Common/DriverHelper.c [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    return device_configurations[device_id];
+}
+
+void*
+get_remote_device_configuration(enum SystemDevice device_id)
+{
+    if(device_id > SYSTEM_DEVICE_NUMBER) {
+        return 0;
+    }
+    enum SystemBus bus_id = device_to_bus_map[device_id];
```
Comment (Glamhoth) [inline src/Common/DriverHelper.c:19]: const

File: src/Common/DriverHelper.c [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+{
+    if(device_id > SYSTEM_DEVICE_NUMBER) {
+        return (void*)0;
+    }
+
+    return device_configurations[device_id];
+}
+
+void*
+get_remote_device_configuration(enum SystemDevice device_id)
```
Comment (Glamhoth) [inline src/Common/DriverHelper.c:14]: const arg

File: src/Common/DriverHelper.c [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+{
+    if(device_id > SYSTEM_DEVICE_NUMBER) {
+        return 0;
```
Comment (Glamhoth) [inline src/Common/DriverHelper.c:17]: cast to `void*`?

File: src/Common/DriverHelper.c [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    enum SystemBus bus_id = device_to_bus_map[device_id];
+    for(int index = 0; index < SYSTEM_DEVICE_NUMBER; ++index) {
+        if(index != device_id) {
+            if(device_to_bus_map[index] == bus_id) {
```
Comment (Glamhoth) [inline src/Common/DriverHelper.c:22]: maybe merge those two `if`s?

File: src/Mocks/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+target_include_directories(Mocks
+  PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_link_libraries(Mocks
+  PRIVATE common_build_options)
+
+add_format_target(Mocks)
+
+add_library(TASTE::Mocks ALIAS Mocks)
```
Comment (Glamhoth) [inline src/Mocks/CMakeLists.txt:14]: Maybe `TASTE::RuntimeMocks` to avoid possible name clash?

File: src/Mocks/system_spec.h [hunk 16]
```diff
@@ -0,0 +1,35 @@
+#ifndef SYSTEM_SPEC_H
+#define SYSTEM_SPEC_H
+
+enum RemoteInterface {
+    INTERFACE_INVALID_ID,
+
+    // Sporadic interfaces from x86_partition
```
Comment (Glamhoth) [inline src/Mocks/system_spec.h:7]: Is it a `TODO`?

File: src/Packetizer/Packetizer.c [hunk 17]
```diff
@@ -93,7 +93,7 @@ Packetizer_depacketize(const Packetizer* const self,
     assert(destination != NULL);
 
     // Get and check data size
-    size_t receivedDataSize = ((size_t)(packetPointer[4] << 8u) | packetPointer[5]) + 1;
+    size_t receivedDataSize = readPacketDataLength(packetPointer);
```
Comment (Glamhoth) [inline src/Packetizer/Packetizer.c:96]: const

File: src/Packetizer/Packetizer.c [hunk 18]
```diff
@@ -175,3 +182,15 @@ writeCrc(uint8_t* const packetPointer, const size_t dataSize)
     packetPointer[SPACE_PACKET_PRIMARY_HEADER_SIZE + dataSize] = (crc >> 8) & 0xFF;
     packetPointer[SPACE_PACKET_PRIMARY_HEADER_SIZE + dataSize + 1] = crc & 0xFF;
 }
+
+size_t
+readPacketDataLength(uint8_t* const packetPointer)
```
Comment (Glamhoth) [inline src/Packetizer/Packetizer.c:187]: const arg

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/Mocks/CMakeLists.txt (hunk 15 above)
Comment (rphh) [reply src/Mocks/CMakeLists.txt]: good idea. changed.

File: src/Packetizer/Packetizer.c (hunk 18 above)
Comment (rphh) [reply src/Packetizer/Packetizer.c:187]: ??

File: src/Broker/Broker.c (hunk 2 above)
Comment (rphh) [reply src/Broker/Broker.c]: added.

File: src/Broker/Broker.c (hunk 3 above)
Comment (rphh) [reply src/Broker/Broker.c]: added.

File: src/Common/DriverHelper.c (hunk 14 above)
Comment (rphh) [reply src/Common/DriverHelper.c]: ack

File: src/Mocks/system_spec.h (hunk 16 above)
Comment (rphh) [reply src/Mocks/system_spec.h]: no, comment removed

File: src/Packetizer/Packetizer.c (hunk 17 above)
Comment (rphh) [reply src/Packetizer/Packetizer.c]: done

File: src/Packetizer/Packetizer.c (hunk 18 above)
Comment (rphh) [reply src/Packetizer/Packetizer.c]: fixed.

File: src/Broker/Broker.c (hunk 4 above)
Comment (rphh) [reply src/Broker/Broker.c]: done

File: src/Common/DriverHelper.c (hunk 10 above)
Comment (rphh) [reply src/Common/DriverHelper.c]: done

File: src/Common/DriverHelper.c (hunk 12 above)
Comment (rphh) [reply src/Common/DriverHelper.c]: done

File: src/Common/DriverHelper.c (hunk 13 above)
Comment (rphh) [reply src/Common/DriverHelper.c]: done

File: src/Broker/CMakeLists.txt (hunk 5 above)
Comment (rphh) [reply src/Broker/CMakeLists.txt]: done

File: src/Broker/CMakeLists.txt (hunk 6 above)
Comment (rphh) [reply src/Broker/CMakeLists.txt]: fixed

File: src/Common/CMakeLists.txt (hunk 7 above)
Comment (rphh) [reply src/Common/CMakeLists.txt]: fixed

File: src/Common/CMakeLists.txt (hunk 8 above)
Comment (rphh) [reply src/Common/CMakeLists.txt]: changed

File: src/Common/CMakeLists.txt (hunk 9 above)
Comment (rphh) [reply src/Common/CMakeLists.txt]: renamed

File: src/Broker/Broker.c [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    Packetizer_init(&packetizer_data);
+}
+
+void
+Broker_deliver_request(enum RemoteInterface interface, uint8_t* const data, const size_t length)
+{
+    Broker_acquire_lock();
+    memcpy(packetizer_buffer + SPACE_PACKET_PRIMARY_HEADER_SIZE, data, length);
+
+    size_t packet_size = Packetizer_packetize(&packetizer_data,
```
Comment (Glamhoth) [inline src/Broker/Broker.c:28]: const

File: src/Broker/Broker.c (hunk 4 above)
Comment (Glamhoth) [reply src/Broker/Broker.c]: what about the enum?

File: src/Broker/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    PRIVATE     Broker.c
+    PUBLIC      Broker.h)
+
+target_include_directories(Broker
+  PUBLIC      ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_link_libraries(Broker
+  PRIVATE     common_build_options
+              TASTE::RuntimeMocks
+              TASTE::Packetizer)
```
Comment (Glamhoth) [inline src/Broker/CMakeLists.txt:14]: I think `Packetizer` should be marked `PUBLIC` as we expose SpacePacket defines in `Broker.h`

File: src/Broker/Broker.h [hunk 21]
```diff
@@ -0,0 +1,22 @@
+#ifndef BROKER_H
+#define BROKER_H
+
+#include <system_spec.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <Packetizer.h>
```
Comment (Glamhoth) [inline src/Broker/Broker.h:9]: Maybe just include `SpacePacket.h` as we only use two defines here?

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/Broker/Broker.c (hunk 4 above)
Comment (rphh) [reply src/Broker/Broker.c]: Ou, I've added it only in header. Fixed.

File: src/Broker/Broker.c (hunk 19 above)
Comment (rphh) [reply src/Broker/Broker.c]: Added.

File: src/Broker/CMakeLists.txt (hunk 20 above)
Comment (rphh) [reply src/Broker/CMakeLists.txt:14]: Hmm, It is not a problem w kazoo, however, indeed it is exposed so potentialy other targets needs to see includes.

File: src/Broker/Broker.h (hunk 21 above)
Comment (rphh) [reply src/Broker/Broker.h:9]: Yes, this should be enaugh. `Packetizer.h` is will be included in `Broker.c`.

File: src/Broker/CMakeLists.txt (hunk 20 above)
Comment (rphh) [reply src/Broker/CMakeLists.txt]: done

File: src/Broker/Broker.h (hunk 21 above)
Comment (rphh) [reply src/Broker/Broker.h]: done

File: src/Broker/Broker.h [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+#include <system_spec.h>
+#include <request_size.h>
+
+/// @brief  Buffer size to store space packet data.
+#define BROKER_BUFFER_SIZE                                                                                             \
+    (SPACE_PACKET_PRIMARY_HEADER_SIZE + GENERIC_LINUX_BUFFER_SIZE + SPACE_PACKET_ERROR_CONTROL_SIZE)
+
+/** @brief Initialize Broker
+ *
+ * Broker shall be initialized at the before starting threads of system.
```
Comment (Glamhoth) [inline src/Broker/Broker.h:57]: `initialized before starting system's threads`?

File: src/Broker/Broker.h [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+ */
+void Broker_initialize(void);
+
+/** @brief deliver message to remote interface
```
Comment (Glamhoth) [inline src/Broker/Broker.h:61]: Capital letter

File: src/Broker/Broker.h [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+ * This function shall be used to deliver a message (sporadic interface call)
+ * to remote partition. This function is used in code generated using kazoo.
+ *
+ * @param[in]    interface      Target remote interface
+ * @param[in]    data           Pointer to an array with the data
+ * @param[in]    length         Size of the data
+ */
+void Broker_deliver_request(const enum RemoteInterface interface, uint8_t* const data, const size_t length);
+
+/** @brief deliver message from remote interface
```
Comment (Glamhoth) [inline src/Broker/Broker.h:72]: Capital letter

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/Broker/Broker.h (hunk 24 above)
Comment (rphh) [reply src/Broker/Broker.h]: done

File: src/Broker/Broker.h (hunk 23 above)
Comment (rphh) [reply src/Broker/Broker.h]: done

File: src/Broker/Broker.h (hunk 22 above)
Comment (rphh) [reply src/Broker/Broker.h]: done

Comment (Glamhoth) [reply src/Broker/Broker.h]: You forgot to remove `at the`

Comment (rphh) [reply src/Broker/Broker.h]: true. fixed.

**Review by Glamhoth — DISMISSED** [review]

File: src/Broker/Broker.c [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void
+Broker_deliver_request(const enum RemoteInterface interface, uint8_t* const data, const size_t length)
+{
+    Broker_acquire_lock();
+    memcpy(packetizer_buffer + SPACE_PACKET_PRIMARY_HEADER_SIZE, data, length);
+
+    const size_t packet_size = Packetizer_packetize(&packetizer_data,
+                                                    Packetizer_PacketType_Telemetry,
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:53]: always telemetry?

File: src/Broker/Broker.c [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+                                                    0,
+                                                    (uint16_t)interface,
+                                                    packetizer_buffer,
+                                                    SPACE_PACKET_PRIMARY_HEADER_SIZE,
+                                                    length);
+
+    // sent to driver
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:60]: senD?

Comment (Lurkerpas) [reply src/Broker/Broker.c:60]: Also, is this comment necessary?

File: src/Broker/Broker.c [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+    const enum SystemBus bus_id = port_to_bus_map[interface];
+    void* driver_private_data = bus_to_driver_private_data[bus_id];
+    driver_send_function send_function = bus_to_driver_send_function[bus_id];
+    send_function(driver_private_data, packetizer_buffer, packet_size);
+
+    Broker_release_lock();
+}
+
+// this shall be called by driver
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:69]: Would be nice if this looked like this:
"This function shall be called by a driver"

File: src/Broker/Broker.c [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+void
+Broker_receive_packet(uint8_t* const data, const size_t length)
+{
+    Broker_acquire_lock();
+    // mutex
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:74]: Is this comment necessary?

File: src/Broker/Broker.c [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+    uint16_t source;
+    uint16_t destination;
+    size_t data_offset;
+    size_t data_size;
+    int32_t error_code;
+    const bool success = Packetizer_depacketize(&packetizer_data,
+                                                Packetizer_PacketType_Telemetry,
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:81]: Always telemetry?

File: src/Broker/Broker.h [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+#include <stdint.h>
+
+#include <SpacePacket.h>
+
+#include <system_spec.h>
+#include <request_size.h>
+
+/// @brief  Buffer size to store space packet data.
+#define BROKER_BUFFER_SIZE                                                                                             \
+    (SPACE_PACKET_PRIMARY_HEADER_SIZE + GENERIC_LINUX_BUFFER_SIZE + SPACE_PACKET_ERROR_CONTROL_SIZE)
```
Comment (Lurkerpas) [inline src/Broker/Broker.h:53]: GENERIC_LINUX_BUFFER_SIZE - broker is not Linux specific

File: src/Packetizer/Packetizer.c [hunk 31]
```diff
@@ -175,3 +183,15 @@ writeCrc(uint8_t* const packetPointer, const size_t dataSize)
     packetPointer[SPACE_PACKET_PRIMARY_HEADER_SIZE + dataSize] = (crc >> 8) & 0xFF;
     packetPointer[SPACE_PACKET_PRIMARY_HEADER_SIZE + dataSize + 1] = crc & 0xFF;
 }
+
+size_t
+readPacketDataLength(const uint8_t* const packetPointer)
+{
+#ifdef STANDARD_SPACE_PACKET
+    return ((size_t)(packetPointer[4] << 8u) | packetPointer[5]) + 1;
+
```
Comment (Lurkerpas) [inline src/Packetizer/Packetizer.c:192]: Is this empty line necessary?

File: src/Broker/Broker.c (hunk 25 above)
Comment (rphh) [reply src/Broker/Broker.c]: Changed. This is configurable. The default option (if nothing is configured) is still telemetry.

File: src/Broker/Broker.c (hunk 26 above)
Comment (rphh) [reply src/Broker/Broker.c]: comment removed.

File: src/Broker/Broker.c (hunk 27 above)
Comment (rphh) [reply src/Broker/Broker.c]: comment removed.

File: src/Broker/Broker.c (hunk 28 above)
Comment (rphh) [reply src/Broker/Broker.c]: comment removed.

File: src/Broker/Broker.c (hunk 29 above)
Comment (rphh) [reply src/Broker/Broker.c]: Changed. This is configurable. The default option (if nothing is configured) is still telemetry.

File: src/Broker/Broker.h (hunk 30 above)
Comment (rphh) [reply src/Broker/Broker.h]: Changed to GENERIC_PARTITION_BUFFER_SIZE (changed also in kazoo templates)

File: src/Packetizer/Packetizer.c (hunk 31 above)
Comment (rphh) [reply src/Packetizer/Packetizer.c:192]: empty line removed

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**